dockerfiles/theia/Dockerfile

```diff
@@ -8,24 +8,102 @@
 # Contributors:
 #   Red Hat, Inc. - initial API and implementation
 
-FROM centos
+###
+# Builder Image
+#
+FROM centos as builder
 
 ARG GITHUB_TOKEN
-ARG THEIA_VERSION=0.3.14
-
-EXPOSE 3000 3030
+ARG THEIA_VERSION=plugin-id
 
 ENV THEIA_VERSION=${THEIA_VERSION} \
-    USE_LOCAL_GIT=true \
     HOME=/home/theia
 
 WORKDIR ${HOME}
 
-ADD src/ ${HOME}/
-ADD config.yaml ${HOME}/.config/verdaccio/
-ADD supervisord.conf /etc/
+# Install basic software used for checking github API rate limit
+RUN yum install -y epel-release && yum -y install curl jq expect
+
+# define in env variable GITHUB_TOKEN only if it is defined
+# else check if github rate limit is enough, else will abort requiring to set GITHUB_TOKEN value
+ADD src/builder/check-github-limit.sh /tmp/builder/scripts/check-github-limit.sh
+RUN /tmp/builder/scripts/check-github-limit.sh
+
+ADD src/builder/setup-nodejs.sh /tmp/builder/scripts/setup-nodejs.sh
+RUN /tmp/builder/scripts/setup-nodejs.sh
+
+# Install tools required to build theia
+RUN yum install -y gcc-c++ make python git
+
+# Add verdaccio config
+ADD src/config.yaml ${HOME}/.config/verdaccio/
+
+# Build theia from source code and publish artifacts locally
+ADD src/patches /home/theia-build/patches
+ADD src/builder/build-and-publish-theia.sh /tmp/builder/scripts/build-and-publish-theia.sh
+RUN /tmp/builder/scripts/build-and-publish-theia.sh
+
+# Now, add the extensions
+ADD src/add-extensions.js /home/theia-build/add-extensions.js
+ADD src/versions.sh /home/theia-build/versions.sh
+ADD src/resolutions-provider.js /home/theia-build/resolutions-provider.js
+
+ADD src/extensions.json ${HOME}/extensions.json
+ADD src/package.json ${HOME}/package.json
+ADD src/builder/add-extensions.sh /tmp/builder/scripts/add-extensions.sh
+RUN /tmp/builder/scripts/add-extensions.sh
+
+# Assemble theia
+ADD src/builder/assemble-theia.sh /tmp/builder/scripts/assemble-theia.sh
+RUN /tmp/builder/scripts/assemble-theia.sh
+
+# change permissions
+RUN find ${HOME} -exec sh -c "chgrp 0 {}; chmod g+rwX {}" \;
+
+# Clean
+RUN yarn cache clean
```

Comment:
I am not sure that we need "yarn cache clean", because this command clean up yarn cache folder for "builder image", but we don't use this for "runtime image".

Reply:
cache is in $HOME/theia folder and we copy that folder to the runtime image.

I would say that if cache is stored outside of this directory then we could avoid to clean it

Reply:
I will try to use YARN_CACHE_FOLDER later (so defining the env var only in the `build` image)